default trackers

stefantalpalaru opened the pull request:
Some torrents benefit greatly from a default list of trackers that speeds up the search for peers. This patch implements that (only for public torrents) for the daemon, web interface, GTK and Qt clients.

The default trackers will be added to new public torrents and to existing ones, after a reload.

The concept comes from this Deluge plugin: https://github.com/stefantalpalaru/deluge-default-trackers

dotnetCarpenter commented:
@stefantalpalaru Is it possible to edit the list? Similar to how it works now:
![Adding a tracker](https://cloud.githubusercontent.com/assets/2473545/25580709/6e981b2c-2e59-11e7-8fee-f960cda918ed.png)

In general it's valuable to remove non-answering (dead) trackers and add trackers that are not found by `tor->session->defaultTrackers`.

*note: cross posted from [Support for µTorrent compatible list](https://github.com/transmission/transmission/issues/237#issuecomment-308434175) because it really belongs here*

stefantalpalaru commented:
Of course it's possible. What you see in the Preferences window is an editable text field. It comes empty and you add the default trackers yourself.

The only catch is that the changes only apply for future torrents (or existing torrents after a restart, but only for adding new default trackers).

thibaultmol commented:
Please accept this pull request @charlesk :-)

jmaris commented:
Has this not yet been merged ?

imanushin commented:
@jordanl, @mikedld and @jre could you please approve or reject PR of @stefantalpalaru? Or possible write any comments regarding this change?

zero77 commented:
Is this possible in the web-ui or in the settings.json file.

stefantalpalaru commented:
> Is this possible in the web-ui or in the settings.json file.

Yes, both are supported.

zero77 commented:
@stefantalpalaru 
Thanks
Also, I may have missed it but, I think the "Editing Configuration Files" in the wiki needs up dating, to show how this is possible.

stefantalpalaru commented:
They need to merge my pull request first.

zero77 commented:
O, good point I missed that.

psycholyzern commented:
any news? I have a torrents with only a few trackers that I access via web-ui. I hope that I can add more trackers via web-ui.

chigix commented:
I'm waiting for this release as well

Talent30 commented:
> I'm waiting for this release as well

Same

ckerr commented on libtransmission/torrent.c:
```diff
@@ -1131,6 +1132,72 @@ static tr_parse_result torrentParseImpl(tr_ctor const* ctor, tr_info* setmeInfo,
+            if (duplicate)
+            {
+                continue;
+            }
```
I feel like this block duplicates the helper function `findAnnounceUrl()`. Maybe that utility could be moved from `rpcimpl.c` to `torrent.[ch]` so that it could be reused here?

stefantalpalaru replied:
Yes, but the function name is misleading in this context, rendering the code less readable.

ckerr commented on libtransmission/torrent.c:
```diff
@@ -1131,6 +1132,72 @@ static tr_parse_result torrentParseImpl(tr_ctor const* ctor, tr_info* setmeInfo,
+
+    if (changed)
+    {
+        tr_torrentSetAnnounceList(tor, trackers, numExistingTrackers);
```
Need to handle what happens if `tr_torrentSetAnnounceList()` returns `false` to indicate failure

stefantalpalaru replied:
I'm already using `tr_urlIsValidTracker()`, so `tr_torrentSetAnnounceList()` has no reason to fail.

ckerr commented on libtransmission/session.c:
```diff
@@ -2359,6 +2366,55 @@ int tr_sessionGetCacheLimit_MB(tr_session const* session)
+        }
+
+        walk += len + 1;
+    }
```
There's a similar block of code in `tr_rpcSetList()` and I like this implementation better.

I wonder if it would make sense to move this code into `utils.[ch]` as a `tr_strsplit(const char* str, const char* delimeters, tr_list** list)` where it could be use by both this code and `tr_rpcSetList()`?

ckerr commented on libtransmission/session.h:
```diff
@@ -222,6 +222,9 @@ struct tr_session
     struct tr_bindinfo* public_ipv6;
+
+    struct tr_list* defaultTrackers;
+    char* defaultTrackersStr;
```
These are leaked; need to be cleaned up at the end of `tr_sessionClose()` in `session.c`

stefantalpalaru replied:
Done.

ckerr commented on libtransmission/session.c:
```diff
@@ -2359,6 +2366,55 @@ int tr_sessionGetCacheLimit_MB(tr_session const* session)
+{
+    TR_ASSERT(tr_isSession(session));
+
+    return session->defaultTrackersStr ? session->defaultTrackersStr : "";
```
The other `char const* tr_sessionGetFoo()` string accessors return `NULL` if there's not a string.

Is there an exceptional need for this function to behave differently?

stefantalpalaru replied:
Changed to NULL.

ckerr commented on README.md:
```diff
@@ -29,6 +29,7 @@ For a more detailed description, and dependencies, visit: https://github.com/tra
     $ tar xf transmission-2.92.tar.xz
     $ cd transmission-2.92
     $ mkdir build
+    $ cd build
```
this seems to be fixed in master already (but that's probably because this PR has sat unreviewed in `tr/tr` for so long)

ckerr commented on gtk/tr-prefs.c:
```diff
@@ -197,6 +197,45 @@ static GtkWidget* new_entry(tr_quark const key, gpointer core)
+    tr_quark const key = GPOINTER_TO_INT(g_object_get_data(G_OBJECT(buffer), PREF_KEY));
+    GtkTextIter start, end;
+    gtk_text_buffer_get_start_iter(buffer, &start);
+    gtk_text_buffer_get_end_iter(buffer, &end);
```
_(minor)_ `gtk_text_buffer_get_bounds(buffer, &start, &end);` would make more sense here

stefantalpalaru replied:
Changed.

ckerr commented on gtk/tr-prefs.c:
```diff
@@ -197,6 +197,45 @@ static GtkWidget* new_entry(tr_quark const key, gpointer core)
+
+    /* key and signal */
+    g_object_set_data(G_OBJECT(buffer), PREF_KEY, GINT_TO_POINTER(key));
+    g_signal_connect(buffer, "changed", G_CALLBACK(text_buffer_changed_cb), core);
```
I _think_ this is OK but my GTK+ is slightly rusty; can you confirm this event isn't fired off for every keystroke?

stefantalpalaru replied:
https://developer.gnome.org/gtk3/stable/GtkEditable.html#GtkEditable-changed :

> The ::changed signal is emitted at the end of a single user-visible operation on the contents of the GtkEditable.
> 
> E.g., a paste operation that replaces the contents of the selection will cause only one signal emission (even though it is implemented by first deleting the selection, then inserting the new content, and may cause multiple ::notify::text signals to be emitted).
>

ckerr commented on libtransmission/torrent.c:
```diff
@@ -1131,6 +1132,72 @@ static tr_parse_result torrentParseImpl(tr_ctor const* ctor, tr_info* setmeInfo,
+    // add the new ones
+    for (l=tor->session->defaultTrackers; l!=NULL; l=l->next)
+    {
+        url = (char*)l->data;
```
the declaration for `url` could be moved to here, narrowing its scope. Also since this is passed walking through the session array it should be a const IMO. Also since `l->data` is a `void*` I don't think we need to explicitly cast it? So I think this would work:

`char const* url = l->data;`

stefantalpalaru replied:
Changed.

ckerr commented on libtransmission/torrent.c:
```diff
@@ -1131,6 +1132,72 @@ static tr_parse_result torrentParseImpl(tr_ctor const* ctor, tr_info* setmeInfo,
+    bool duplicate;
+    bool changed = false;
+
+    if (!numNewTrackers || tor->info.isPrivate)
```
it's slightly cleaner to use the `tr_torrentIsPrivate(tor)` helper for the second test

stefantalpalaru replied:
Changed.

ckerr commented on libtransmission/torrent.c:
```diff
@@ -1131,6 +1132,72 @@ static tr_parse_result torrentParseImpl(tr_ctor const* ctor, tr_info* setmeInfo,
+    int numExistingTrackers = tor->info.trackerCount;
+    int numNewTrackers = tr_list_size(tor->session->defaultTrackers);
+    char* url;
+    bool duplicate;
```
`url` and `duplicate` can be declared further down, inside the loop, to narrow the variables' scopes.

stefantalpalaru replied:
Done.

ckerr commented on libtransmission/torrent.c:
```diff
@@ -1131,6 +1132,72 @@ static tr_parse_result torrentParseImpl(tr_ctor const* ctor, tr_info* setmeInfo,
+        return;
+    }
+
+    // copy existing trackers
```
it's slightly more readable to initialize `tier` to `-1` here, where we're using it, instead of at the beginning of the function

stefantalpalaru replied:
Changed.

Codingale commented:
@stefantalpalaru could you re-visit this and update it and mention someone once you do? I'd really love this feature and would implement it myself but I'm clueless when it comes to c and those changes requested above.

stefantalpalaru commented:
@Codingale - I rebased my PR, addressed most review requests and cleaned up the syntax a bit more.

By the way, I've been using this patch for the last couple of years, packaged into a Gentoo ebuild: https://github.com/stefantalpalaru/gentoo-overlay/tree/master/net-p2p/transmission

uyu423 commented:
I think this Pull Request is a really necessary feature for Transmission Client. Anyone know about the progress of this issue?

stefantalpalaru commented:
Rebased on top of the master branch.

dotnetCarpenter commented:
@mikedld Could you please merge this in? A lot of people wants this for years! The only changes requested by @ckerr, that are not addressed, seems to be outside the scope of this PR and has to do with improving other areas of Transmission.

stefantalpalaru commented:
Rebased on top of the master branch.

bassmadrigal commented:
I love this feature and I really hope the accept the pull request. Thanks for keeping it updated for those of us who merge it ourselves.

dotnetCarpenter commented:
@ckerr please merge this PR. A lot of transmission users want this feature, including me.

RobFog commented:
Another rebase is necessary.

stefantalpalaru commented:
Rebased on "master".

texasbruce commented:
transmission is literally unusable right now with all those default trackers being inactive. This PR should be merged long ago and its been 4 years now.

texasbruce commented:
transmission is literally unusable right now with all those default trackers being inactive. This PR should be merged long ago and its been 4 years now.

@stefantalpalaru If you can, could you maybe build the binary releases? I'll use your version

stefantalpalaru commented:
> @stefantalpalaru If you can, could you maybe build the binary releases? I'll use your version

Oh, man, it's been four years already?

I can only release binaries if I make a proper fork, with a new name, and maintaining that is just too much work. I rebased the PR instead ;-)

kubrickfr commented:
Hello maintainers! 
Is there a reason why this is not merged? IE, is there a reason why I should not use this fork?

Codingale commented:
> Is there a reason why this is not merged? IE, is there a reason why I should not use this fork?

Just someone who's still subscribed to this thread, but from what I can tell. Not really, just probably a lot of workload on the maintainers of the repo. That and the automated checks seem to be broken as two haven't actually said anything since March.

KoalaV2 commented:
Why has this not been merged yet? This is one of the primary reasons I'm still not using transmission. Would be very very useful.

muzuiget commented:
I share my workaround, we can use the Transmission HTTP interface, enable it in the Transmission setting panel "Remote".

Then write a simple Python script to send the tracker urls:

```
$ cat main.py
#!/usr/bin/env python3

import sys
from subprocess import run

def main():
    if len(sys.argv) != 2:
        run(['transmission-remote', '-l'], check=True)
        return
    t = sys.argv[1]
    path = 'trackers.txt'
    for td in map(lambda x: x.strip(), open(path, 'r').readlines()):
        if td != '':
            run(['transmission-remote', '-t', t, '-td', td], check=False)

if __name__ == '__main__':
    main()
```

Prepare the trackers file:

```
$ cat trackers.txt
udp://tracker1.example.com:80/announce
udp://tracker2.example.com:80/announce
udp://tracker2.example.com:80/announce
```

First, get the torrent ID：

```
$ ./main.py 
    ID   Done       Have  ETA           Up    Down  Ratio  Status       Name
     1   100%    4.62 GB  Done         0.0     0.0    0.0  Stopped      Example Torrent 1
     2   100%    8.00 GB  Done         0.0     0.0    0.1  Stopped      Example Torrent 2
Sum:            12.62 GB               0.0     0.0
```

Then execute the script with the torrent ID argument:

```
$ ./main.py 1
localhost:9091/transmission/rpc/ responded: "success"
localhost:9091/transmission/rpc/ responded: "success"
localhost:9091/transmission/rpc/ responded: "success"
```

Check the result int the Transmission GUI.

erhan- commented:
Can you please merge this after so many years?

bison92 commented:
after 5 years, I'm downloading uTorrent. Cya!

KoalaV2 commented:
Incredibly stupid for this to not have been merged yet.

ckerr reviewed (changes requested):
In order for this to be reviewed & merged now, this PR needs to be rebased/updated to match the C++ification of daemon / libtransmission / transmission-gtk.

If you do update this, I'll try to not let the PR set for another 5 years before doing a followup review... :snail:

stefantalpalaru commented:
Patch translated to C++ and gtkmm.

Every day we stray further away from C's light...

ckerr commented on libtransmission/rpcimpl.cc:
```diff
@@ -1789,6 +1789,11 @@ static char const* sessionSet(
         tr_sessionSetQueueStalledEnabled(session, boolVal);
     }
 
+    if (tr_variantDictFindStrView(args_in, TR_KEY_default_trackers, &sv))
```
I'm a little concerned about the idea of passing a tracker list as a single string in settings, e.g. when a users enters this into a textfield and presses return to make a newline, how do we serialize that in settings.json? JSON doesn't handle newlines embedded in strings.

OT1H we could possibly store this in settings.json as `[ [ "tier1url1", "tier1url2" ], [ "tier2url1", tier2url2" ] ... ]`. OTOH I don't actually know if the code would work, since right now we largely use settings.json as a glorified ini file with single-depth key/value pairs. Refactoring the settings code would be out-of-scope for this PR.

Just brainstorming -- what if these were stored in a hardcoded `${session->config_dir}/default-trackers.txt`, similar to how blocklists are stored now? That would sidestep the issue of how to encode this in settings.

bassmadrigal replied:
> Just brainstorming -- what if these were stored in a hardcoded ${session->config_dir}/default-trackers.txt, similar to how blocklists are stored now?

I know I'm just an end user, but I love this idea. I remember struggling trying to figure out how to add it properly when I first cherry picked the this PR into my local build. I was using transmission-daemon and ended up loading transmission-gtk to add the URLs through the GUI settings and then looking at the settings file to adjust it on my daemon settings.

Having it as a text file would make updates super easy too.

stefantalpalaru replied:
> how do we serialize that in settings.json? JSON doesn't handle newlines embedded in strings.

Look at "~/.config/transmission/settings.json" - newlines are serialised as "\n" and it works just fine.

I've been using this patch for almost five years. It works as advertised.

ckerr commented on libtransmission/torrent.cc:
```diff
@@ -817,6 +817,53 @@ static void torrentInit(tr_torrent* tor, tr_ctor const* ctor)
+    }
+    /* tell the announcer to reload this torrent's tracker list */
+    tr_announcerResetTorrent(tor->session->announcer, tor);
+    tor->announceList().save(tor->torrentFile());
```
So once public a torrent is added, these URLs are "sticky" and will persist even if the user decides they want to change the global announce list to a different set?

I think it would be better if these were applied dynamically instead of being saved in the torrent.

stefantalpalaru replied:
Yeah, you're right. Saving removed.

ckerr commented on libtransmission/torrent.cc:
```diff
@@ -817,6 +817,53 @@ static void torrentInit(tr_torrent* tor, tr_ctor const* ctor)
+    std::list<std::string> trackerURLs = {};
+    int numExistingTrackers = tr_torrentTrackerCount(tor);
+    int numNewTrackers = tor->session->defaultTrackersList.size();
+
+    if (!numNewTrackers || tor->isPrivate())
+    {
+        return;
+    }
+
+    // copy existing tracker URLs
+    for (int i = 0; i < numExistingTrackers; ++i)
+    {
+        auto tracker = tr_torrentTracker(tor, i);
+        trackerURLs.push_back(tracker.announce);
+    }
+
+    // add the new ones
+    for (std::string_view url : tor->session->defaultTrackersList)
+    {
+        if (tr_urlIsValidTracker(url))
+        {
+            // check for duplicates
+            bool duplicate = false;
+            for (auto trackerURL : trackerURLs)
+            {
+                if (trackerURL == url)
+                {
+                    duplicate = true;
+                    break;
+                }
+            }
+
+            if (duplicate)
+            {
+                continue;
+            }
+
+            tor->announceList().add(url);
+        }
+    }
```
`tr_announce_list.add()` already ignores duplicates, so no need to check them here

```suggestion
    if (tor->isPrivate())
    {
        return;
    }
    
    for (auto const& url : tor->session->defaultTrackersList)
    {
            tor->announceList().add(url);
    }
```

ckerr commented on qt/PrefsDialog.cc:
```diff
@@ -295,6 +307,22 @@ void PrefsDialog::pathChanged(QString const& path)
+        if (plain_text_edit->document()->isModified())
+        {
+            prefs_.set(pref_widget.getPrefKey(), plain_text_edit->toPlainText());
+            // we avoid using setPref() because the included refreshPref() call would reset the cursor while we're editing
```
Maybe connect to `focusChanged`? I don't have the docs in front of me so I don't know if it's doable, but if so it seems like it would be simpler

stefantalpalaru replied:
If it ain't broken...

ckerr replied:
Does this event get fired every time a character is changed in the text edit, or once when the user is done? E.g. what happens if someone physically types the URLs into the text edit?

stefantalpalaru replied:
"[QTextDocument](https://doc.qt.io/qt-5/qtextdocument.html) emits a [textChanged](https://doc.qt.io/qt-5/qplaintextedit.html#textChanged)() signal if the text changes and it also provides a isModified() function which will return true if the text has been modified since it was either loaded or since the last call to setModified with false as argument." - https://doc.qt.io/qt-5/qplaintextedit.html

So yes, modifying URLs in that text box is working. There might be some internal Qt lag between typing and triggering the signal, to avoid triggering it too often.

ckerr replied:
> So yes, modifying URLs in that text box is working.

My concern is not that zero modifications are made, but that when someone starts typing, we will flood tr_session with hundreds of tracker rebuilds.

Fixed in https://github.com/transmission/transmission/pull/2668/commits/3780f90ef554e21504e5ac2ddfe79ff4b602c68c#diff-3ebbc2c0d8da18b7c49534751edd57bb14351013218ef4e192e5287d6e0c9033R263 by comparing the text contents on focus-in + focus-out and modifying the prefs iff the two differ

ckerr commented on gtk/PrefsDialog.cc:
```diff
@@ -1051,6 +1084,12 @@ Gtk::Widget* PrefsDialog::Impl::networkPage()
+    t->add_section_title(row, _("Default Trackers"));
+    auto tv = new_text_view(TR_KEY_default_trackers, core_);
+    tv->set_tooltip_text(
+        _("a list of default trackers to be added to new public torrents (and existing ones, after a reload)"));
```
...if we just merge these at run-time instead of saving them into newly-added torrents' announce lists, then changes made here could take effect live in libtransmission by walking through public torrents and re-merging the announce  lists from the .torrents with this

stefantalpalaru replied:
Done.

stefantalpalaru commented on libtransmission/session.cc:
```diff
@@ -2244,6 +2252,61 @@ int tr_sessionGetCacheLimit_MB(tr_session const* session)
+    /* build the new list entries */
+    auto urlStart = std::string::npos;
+    auto urlEnd = std::string::npos;
+    char const* delimiters = " ,;\r\n\t";
```
Done.

stefantalpalaru commented on libtransmission/session.h:
```diff
@@ -398,6 +409,7 @@ struct tr_session
     std::array<std::string, TR_SCRIPT_N_TYPES> scripts_;
     std::string blocklist_url_;
     std::string download_dir_;
+    std::string default_trackers_;
```
Renamed. I'd rather keep the original string, in case parsing gets smarter in the future.

ckerr replied:
I don't understand what this means. Do we ever re-parse the string? It looks like we just parse it once when it's passed in to the session

stefantalpalaru replied:
We parse it every time a new session starts, so when the program starts.
The original string also exists offline, while the resulting linked list of individual URLs only exists at runtime.

stefantalpalaru commented on libtransmission/session.cc:
```diff
@@ -2244,6 +2252,61 @@ int tr_sessionGetCacheLimit_MB(tr_session const* session)
+
+    /* build the new list entries */
+    auto urlStart = std::string::npos;
+    auto urlEnd = std::string::npos;
```
Done.

stefantalpalaru commented on libtransmission/session.cc:
```diff
@@ -2244,6 +2252,61 @@ int tr_sessionGetCacheLimit_MB(tr_session const* session)
 ****
 ***/
 
+void tr_session::setDefaultTrackers(std::string_view trackers)
```
I had automated tracker tiers, a few weeks ago, until `tor->announceList().add()` lost its ability to pass a "tier" parameter.

That parser of yours is way too conservative for random URL lists on the Internet.

ckerr replied:
> That parser of yours is way too conservative for random URL lists on the Internet.

It's currently being used to parse utorrent-style tracker lists, so if there are URLs that it can't parsed, that needs to be fixed. Do you have examples of lists that break it?

stefantalpalaru replied:
Can't find any. Just trying to follow https://en.wikipedia.org/wiki/Robustness_principle

ckerr replied:
If you can't find anything that breaks it, how is it too conservative for use here?

stefantalpalaru replied:
I don't really care how it's done, as long as I stop wasting time maintaining this patch, while getting to keep using the functionality.

So feel free to improve it after merging, any way you see fit. I only want default trackers for public torrents in Transmission and a nice UI to manage them. The rest are details.

ckerr replied:
Well, that answers the question of intent.

ckerr replied:
> I had automated tracker tiers, a few weeks ago, until `tor->announceList().add()` lost its ability to pass a "tier" parameter.

It didn't though? https://github.com/transmission/transmission/blob/02b6cc76d197ae74aed60f2f824385e9cfa619f4/libtransmission/announce-list.h#L100

stefantalpalaru replied:
Oh, you just switched param order.

Then why did the tier disappear from what I was modelling from?

https://github.com/transmission/transmission/commit/52f6e80a61ce7340c6e9a60689e345975f02bf8d#diff-6b3564ae830bc5febeda0f04be5f5043ad7a0880df7f034d5538db0a2144ea93L1056

I see that `tr_announce_list::nextTier()` is still there, so we can bring the one-tier-per-tracker-URL scheme back, if you want it.

ckerr replied:
> Then why did the tier disappear from what I was modelling from?

When a tier isn't specified, `nextTier()` is used as the default. That's why it has to go last, since the argument has a default value

ckerr replied:
> I see that `tr_announce_list::nextTier()` is still there, so we can bring the one-tier-per-tracker-URL scheme back, if you want it.

Changing it wouldn't be useful now. When you wrote afternoon that I should do the rest of the details myself, I made a new branch using https://github.com/transmission/transmission/pull/229.patch as a starting point & keeping your authorship in the commit history.

I don't know how you feel about this patch or the project's maintainers after so long, but FWIW I appreciate the work you've done on this.  **Very few** PR submitters take the time to make a new feature work on multiple clients. And add updating it to C++ / gtkmm as well. You did a good job and I appreciate it.

stefantalpalaru commented on libtransmission/session.cc:
```diff
@@ -2244,6 +2252,61 @@ int tr_sessionGetCacheLimit_MB(tr_session const* session)
+    auto fragment = default_trackers_;
+    while ((urlStart = fragment.find_first_not_of(delimiters)) != std::string::npos)
+    {
+        urlEnd = fragment.find_first_of(delimiters, urlStart);
```
No, because the search starts at the "urlStart" position: https://en.cppreference.com/w/cpp/string/basic_string/find_first_of

stefantalpalaru commented on gtk/PrefsDialog.cc:
```diff
@@ -176,6 +176,39 @@ Gtk::Entry* new_entry(tr_quark const key, Glib::RefPtr<Session> const& core)
+
+Gtk::Widget* new_text_view(tr_quark const key, Glib::RefPtr<Session> const& core)
+{
+    auto* scroll = Gtk::make_managed<Gtk::ScrolledWindow>();
```
Done.

stefantalpalaru commented on gtk/PrefsDialog.cc:
```diff
@@ -176,6 +176,39 @@ Gtk::Entry* new_entry(tr_quark const key, Glib::RefPtr<Session> const& core)
+    if (!value.empty())
+    {
+        buffer->set_text(value);
+    }
```
Done.

stefantalpalaru commented on libtransmission/transmission.h:
```diff
@@ -392,6 +392,10 @@ bool tr_sessionIsRPCPasswordEnabled(tr_session const* session);
 
+void tr_sessionSetDefaultTrackers(tr_session* session, char const* trackers);
+
+char const* tr_sessionGetDefaultTrackers(tr_session const* session);
```
Removed.

stefantalpalaru commented on libtransmission/session.cc:
```diff
@@ -407,10 +408,12 @@ void tr_sessionGetSettings(tr_session const* s, tr_variant* d)
     tr_variantDictAddBool(d, TR_KEY_utp_enabled, s->isUTPEnabled);
     tr_variantDictAddBool(d, TR_KEY_lpd_enabled, s->isLPDEnabled);
     tr_variantDictAddStr(d, TR_KEY_download_dir, tr_sessionGetDownloadDir(s));
+    tr_variantDictAddStr(d, TR_KEY_default_trackers, s->defaultTrackers());
```
Fixed.

ckerr commented on daemon/daemon.cc:
```diff
@@ -103,6 +103,7 @@ static auto constexpr Options = std::array<tr_option, 43>{
       { 941, "incomplete-dir", "Where to store new torrents until they're complete", nullptr, true, "<directory>" },
       { 942, "no-incomplete-dir", "Don't store incomplete torrents in a different location", nullptr, false, nullptr },
       { 'd', "dump-settings", "Dump the settings and exit", "d", false, nullptr },
+      { 943, "default-trackers", "Default trackers to be automatically added to public torrents", nullptr, true, "<list>" },
```
_(nit)_ Update the description here slightly to match the non-sticky implementation

```suggestion
      { 943, "default-trackers", "Trackers for public torrents to use automatically", nullptr, true, "<list>" },
```

ckerr commented on gtk/PrefsDialog.cc:
```diff
@@ -176,6 +176,35 @@ Gtk::Entry* new_entry(tr_quark const key, Glib::RefPtr<Session> const& core)
+{
+    Gtk::TextBuffer::iterator start, end;
+    buffer->get_bounds(start, end);
+    Glib::ustring value = buffer->get_text(start, end, FALSE);
+
+    core->set_pref(key, value);
```
_(nit)_ temporary is only used once, so it could be replaced with

```suggestion
    core->set_pref(key, buffer->get_text(start, end, FALSE));
```

Which might give us some C++-move magic

ckerr commented on gtk/PrefsDialog.cc:
```diff
@@ -1051,6 +1080,12 @@ Gtk::Widget* PrefsDialog::Impl::networkPage()
+    t->add_section_title(row, _("Default Trackers"));
+    auto tv = new_text_view(TR_KEY_default_trackers, core_);
+    tv->set_tooltip_text(
+        _("a list of default trackers to be added to new public torrents (and existing ones, after restarting)"));
```
```suggestion
        _("Trackers for public torrents to use automatically"));
```

(doesn't have to be this phrasing word-for-word, but something consistent with the other UI and the daemon CLI)

ckerr commented:
@stefantalpalaru I know it's been a minute since you wrote this, but do you remember what the block/unblock signal was for in https://github.com/stefantalpalaru/transmission/commit/fc371025e7c333b1d50d587cf74f2cfcf9c02eae#diff-3ebbc2c0d8da18b7c49534751edd57bb14351013218ef4e192e5287d6e0c9033R803 ?

It doesn't seem directly related to this feature, so I'm wondering if this was a side quest to fix a bug?

ckerr commented:
The followup PR is up now at https://github.com/transmission/transmission/pull/2668. @stefantalpalaru given your previous comments I'm not sure how much further you want to work on this, but _if you want to_, PTAL and feel free to code review.

stefantalpalaru commented:
> do you remember what the block/unblock signal was for

"Sets the model item index to be the current item, and emits [currentChanged](https://doc.qt.io/qt-5/qitemselectionmodel.html#currentChanged)()." - https://doc.qt.io/qt-5/qitemselectionmodel.html#setCurrentIndex

It wasn't enough, as you'll eventually find out, so I did this instead: https://github.com/stefantalpalaru/transmission/commit/fc371025e7c333b1d50d587cf74f2cfcf9c02eae#diff-3ebbc2c0d8da18b7c49534751edd57bb14351013218ef4e192e5287d6e0c9033R317

But I left the signal disabling in place, because it's good hygiene. Filling in UI widgets from config settings should not trigger "currentChanged" signals because that's not the kind of change we want to intercept and save.
